File: java/src/hackerrank/FairRations.java
package hackerrank;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

// https://www.hackerrank.com/challenges/fair-rations/problem
public class FairRations {

    public static String fairRations(List<Integer> B) {
        var totalOddNrs = B.stream().reduce(0, (acc, b) -> acc = isOdd(b) ? acc+1 : acc);
        if (totalOddNrs == 0)  // only even numbers
            return "0";
        if (isOdd(totalOddNrs))
            return "NO";
        var count = 0;
        for (int i = 0; i < B.size()-1; i++) {
            if (isOdd(B.get(i))) {
                B.set(i+1, B.get(i+1) +1);
                count += 2;
            }
        }
        return String.valueOf(count);
    }

    private static boolean isOdd(int n) {
        return n % 2 != 0;
    }

    @Test
    @DisplayName("Test sample input 0")
    void test0() {
        List<Integer> b = Arrays.asList(2, 3, 4, 5, 6);
        var expected = "4";
        var result = fairRations(b);
        assertEquals(expected, result);
    }

    @Test
    @DisplayName("Test sample input 0")
    void test1() {
        List<Integer> b = Arrays.asList(1, 2);
        var expected = "NO";
        var result = fairRations(b);
        assertEquals(expected, result);
    }

}

File: kotlin/src/hackerrank/fairRations.kt
package hackerrank

import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

// https://www.hackerrank.com/challenges/fair-rations/problem

fun fairRations(B: Array<Int>): String {
    val totalOddNrs = B.count { it.isOdd() }
    if (totalOddNrs == 0)  // only even numbers
        return "0"
    if (totalOddNrs.isOdd() )
        return "NO"
    var count = 0
    for (i in B.indices) {
        if (B[i].isOdd()) {
            B[i] = B[i] + 1
            B[i+1] = B[i+1] +1
            count += 2
        }
    }
    return count.toString()
}

fun Int.isOdd() = this % 2 != 0


internal class FairRationsTest {

    @Test
    @DisplayName("Test sample input 0")
    fun test0() {
        val b: Array<Int> = arrayOf(2, 3, 4, 5, 6)
        val expected = "4"
        val result: String = fairRations(b)
        assertEquals(expected, result)
    }

    @Test
    @DisplayName("Test sample input 1")
    fun test1() {
        val b: Array<Int> = arrayOf(1, 2)
        val expected = "NO"
        val result: String = fairRations(b)
        assertEquals(expected, result)
    }

}